react-router: Add union to component properties

sbuzonas opened the pull request:
This is a different approach to #17086 it solves the issue in #13689

All components that are children of Route are passed the RouteComponentProps, this is typed
as a partial because not all components are router aware, and do not declare/use the props.

Components also extend a generic class that takes properties and state types, these changes
allow for inclusion of the user defined types with a union of the RouteComponentProps provided
by the react-router library.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/ReactTraining/react-router/blob/v4.1.1/packages/react-router/modules/Route.js#L101-L110>
- [ ] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

DovydasNavickas commented on types/react-router/index.d.ts:
```diff
@@ -62,34 +63,36 @@ export interface RouteComponentProps<P> {
+export interface RouteProps<P> {
   location?: H.Location;
-  component?: React.SFC<RouteComponentProps<any> | undefined> | React.ComponentClass<RouteComponentProps<any> | undefined>;
+  component?: React.SFC<RouteComponentPropsUnion<P> | undefined> | React.ComponentClass<RouteComponentPropsUnion<P> | undefined>;
```
Are you sure `P` from `RouteProps` can be used for component props constraint?
Also, not sure if updating `RouteComponentProps` to be `Partial` is correct.

sbuzonas replied:
The `P` was added specifically to be used there. The `Partial` was intended, but perhaps not the correct approach. 

The `RouteComponentProps<any>` properties are always sent, but most third party components passed to the component property are unaware of them and rely solely on the properties in `P`.

DovydasNavickas commented:
Why did you add empty props object to all `React.Component` usages?..

DovydasNavickas commented:
Come on, guys...
Read these comments and update your definitions.
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17288#issuecomment-311946890
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17288#issuecomment-311952176

RyanCavanaugh commented:
<!--header {"slug":"complaint"} headerend-->
@sbuzonas Please fix the failures indicated in the Travis CI log.
<!--footer bot @ 7/14/2017, 11:08:13 AM footerend-->

sbuzonas commented:
The `React.Component<..., {}>` were left over from the project I was testing these definitions against that didn't have #17288. I've updated them. I've also added `{}` as a default to `RouteProps` which should fix the build failure.

I'll still need to rebase and resolve the conflicts

sbuzonas commented:
Is there a way to run a tests for a single package so I can resolve the failures quicker?  The only thing I saw about testing in the readme was to dts lint a single package.

sbuzonas commented:
The broken test is a result of tsd changes in reactjs/redux#2467

andy-ms commented on types/react-router/index.d.ts:
```diff
@@ -62,9 +63,11 @@ export interface RouteComponentProps<P> {
 }
 
-export interface RouteProps {
+export type RouteComponentPropsUnion<P> = RouteComponentProps<any> & P;
```
Might be better to just inline this.

mhegazy commented:
Thanks for your contribution. This PR has failing CI tests and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.
